src/main/java/uk/gov/companieshouse/officerfiling/api/interceptor/ValidTransactionInterceptor.java

```diff
@@ -0,0 +1,57 @@
+package uk.gov.companieshouse.officerfiling.api.interceptor;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.lang.NonNull;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.HandlerMapping;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.officerfiling.api.service.OfficerFilingService;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Map;
+
+public class ValidTransactionInterceptor implements HandlerInterceptor {
+    @Autowired
+    private Logger logger;
+    @Autowired
+    private OfficerFilingService officerFilingService;
+
+    public ValidTransactionInterceptor(){}
+
+    public ValidTransactionInterceptor(Logger logger, OfficerFilingService officerFilingService) {
+        this.logger = logger;
+        this.officerFilingService = officerFilingService;
+    }
+
+    @Override
+    public boolean preHandle(@NonNull HttpServletRequest request,
+                             @NonNull HttpServletResponse response,
+                             @NonNull Object handler) {
+
+        final Map<String, String> pathVariables = (Map)request.getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
+
+        final String transactionId = pathVariables.get("transactionId");
+        final String filingId = pathVariables.get("filingResourceId");
+
+        // check filing exists
+        final var officerFiling = officerFilingService.get(filingId, transactionId);
+        if (officerFiling.isEmpty()) {
+            logger.errorRequest(request, "Filing resource not found");
+            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+            return false;
+        }
+
+        // check filing id from request matches filing id from transaction
+        var matchingOfficerFiling = officerFiling
+                .filter(filing -> officerFilingService.requestUriContainsFilingSelfLink(request, filing));
+
+        if (matchingOfficerFiling.isPresent()) {
```

Comment:
The filter will cause the optional to return empty in the case of no match, as officer filing here is of type Optional.

The isPresent in the if statement will be false in the case of no match.

The requestUriContainsFilingSelfLink method checks that the request URI contains the contents of the self link as it's prefix.

Reply:
This is more concerned with a nefarious activity, where someone has taken a valid URL and say altered the transaction id and then sent that in.  This would capture that scenario.